cluster_management/src/main/java/com/pinterest/rocksplicator/Spectator.java

```diff
@@ -116,25 +119,34 @@ public static void main(String[] args) throws Exception {
     ));
     CommandLine cmd = processCommandLineArgs(args);
     final String zkConnectString = cmd.getOptionValue(zkServer);
-    final String clusterName = cmd.getOptionValue(cluster);
+    final String[] clusterNamesList = cmd.getOptionValues(clusterNames);
     final String host = cmd.getOptionValue(hostAddress);
     final String port = cmd.getOptionValue(hostPort);
     final String postUrl = cmd.getOptionValue(configPostUrl);
     final String instanceName = host + "_" + port;
 
     LOG.error("Starting spectator with ZK:" + zkConnectString);
-    Spectator spectator= new Spectator(zkConnectString, clusterName, instanceName);
-
     CuratorFramework zkClient = CuratorFrameworkFactory.newClient(zkConnectString, new ExponentialBackoffRetry(1000, 3));
     zkClient.start();
-    InterProcessMutex mutex = new InterProcessMutex(zkClient, getClusterLockPath(clusterName));
-    try (Locker locker = new Locker(mutex)) {
-      spectator.startListener(postUrl);
-      Thread.currentThread().join();
-    } catch (RuntimeException e) {
-      LOG.error("RuntimeException thrown by cluster " + clusterName, e);
-    } catch (Exception e) {
-      LOG.error("Failed to release the mutex for cluster " + clusterName, e);
+
+    int i = 0;
+    while (true) {
+      String currentClusterName = clusterNamesList[i];
+      LOG.error("Trying to get lock for cluster " + currentClusterName);
+      InterProcessMutex mutex = new InterProcessMutex(zkClient, getClusterLockPath(currentClusterName));
+      try (Locker locker = new Locker(mutex, 5, TimeUnit.SECONDS)) {
```

Comment:
so one host will get the lock for the first cluster and start generating config for it, and all other hosts will just be blocked?
Who will be responsible to generate configs for other clusters?

Reply:
this is the old implementation recommended by terry. if another host cannot get some clusters lock after five seconds, it will try some other cluster until it eventually obtains some lock